source/agora/test/BlockRewards.d

```diff
@@ -0,0 +1,85 @@
+/*******************************************************************************
+
+    Checks the rewards are correct in the payout blocks
+
+    Copyright:
+        Copyright (c) 2019-2021 BOSAGORA Foundation
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.BlockRewards;
+
+version (unittest):
+
+import agora.api.Validator;
+import agora.common.Amount;
+import agora.consensus.data.Transaction;
+import agora.test.Base;
+
+/// BlockRewards test with empty blocks so that fees are not added to the payouts
+unittest
+{
+    const int period = 4;
+    TestConf conf;
+    conf.consensus.payout_period = period;
+    auto network = makeTestNetwork!TestAPIManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+    auto node_1 = nodes[0];
+
+    // Create blocks until third payout block
+    auto target_height = Height(GenesisValidatorCycle);
+    network.generateBlocks(target_height, true);
+    network.assertSameBlocks(target_height);
+
+    assert(node_1.getBlocksFrom(Height(period), 1).front.txs.empty); // No Coinbase until second payout period ends
+
+    // year 1: coins per 5 secs: commons = 50 validators = 27
+    iota(2 * period, GenesisValidatorCycle + 1, period).each!((height)
+    {
+        auto cb_tx = node_1.getBlocksFrom(Height(height), 1).front.txs.front;
+        cb_tx.outputs.each!(o => assert(o.type == OutputType.Coinbase));
+        assert(cb_tx.outputs.walkLength == 7);
+        assert(cb_tx.outputs.map!(o => o.value).reduce!((a, b) => a + b) == (50.coins + 27.coins) * (600 / 5) * period);
+    });
+}
+
+/// BlockRewards test with blocks with fees added to the payouts
+unittest
+{
+    const int period = 4;
+    TestConf conf;
+    conf.consensus.payout_period = period;
+    auto network = makeTestNetwork!TestAPIManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+    auto node_1 = nodes[0];
+
+    // Create blocks until third payout block
+    auto target_height = Height(GenesisValidatorCycle);
+    network.generateBlocks(target_height);
+    network.assertSameBlocks(target_height);
+
+    // year 1: coins per 5 secs: commons = 50 validators = 27
+    iota(2 * period, GenesisValidatorCycle + 1, period).each!((height)
+    {
+        auto cb_tx = node_1.getBlocksFrom(Height(height), 1).front.txs.filter!(tx => tx.isCoinbase).front;
+        assert(cb_tx.outputs.map!(o => o.value).reduce!((a, b) => a + b) > (50.coins + 27.coins) * (600 / 5) * period);
+    });
+}
+
+/// TODO: test with a validator slashed
+
+/// TODO: test with some missing signatures
```

Comment:
I guess this is still a draft then ? Especially in light of #2503

Reply:
Yes, I will add the tests today